Inc13

icyrogue opened the pull request.

icyrogue committed:
started inc9

icyrogue committed:
fixed status code inc9

icyrogue committed:
fixed header inc9

icyrogue committed:
fixed header v2 inc9

icyrogue committed:
fixed header v3 inc9

icyrogue committed:
added db inc10

icyrogue committed:
started inc10

icyrogue committed:
Merge pull request #9 from icyrogue/inc10

Inc11

icyrogue committed:
started inc11

icyrogue committed:
started inc11

icyrogue committed:
started 21 dawdinc11

icyrogue committed:
started 21 dawdinc11

icyrogue committed:
started 21 dawdincdfssesfef11

icyrogue committed:
started 21 dawdincdfssesfeghfthtfhf11

icyrogue committed:
adwadawdwadaw

icyrogue committed:
adwadawdwadawadwdaw

icyrogue committed:
adwadawdwadawadwdasfsefefw

icyrogue committed:
adwadawdwadawadwdasfsefefdawdaawdw

icyrogue committed:
adwadawdwadawadwdasfsefefdawdaawddadwadw

icyrogue committed:
adwadawdwadawadwdasfsefefdawdaawddadwadw

icyrogue committed:
adwadawdwadawadwdasfsefefdawdaawddadwadgdfgrw

icyrogue committed:
why?

icyrogue committed:
why? v2

icyrogue committed:
why? v3

icyrogue committed:
inc11 done I guess

icyrogue committed:
inc11 done I guess

icyrogue committed:
Merge branch 'inc10' into inc11

icyrogue committed:
inc12

inc12???

inc12IbrokeGitHub

icyrogue committed:
inc13

inc13

inc13 v2

inc13 v4

fixed tests inc13

fixed vet inc13

fixed vet inc13

bambruysk commented on internal/usermanager/usermanager.go:
```diff
@@ -0,0 +1,89 @@
+		fmt.Println(mu.users)
+		return errors.New("no such user")
+	}
+	usr.urls[id] = url
```
Опять мапа без мьютексов?))

bambruysk commented on internal/urlstorage/urlstorage.go:
```diff
@@ -131,3 +135,13 @@ func recoverData(flPath string) (map[string]string, error) {
+
+
+func (st *storage) BulkAdd(data []jsonmodels.JSONBulkInput) error {
+
```
Почему этот метод возвращает пусто?

icyrogue replied:
Я не знал, стоит ли реализовывать метод множественного добавления, если запуск вне базы данных, сейчас что нибудь придумаю

bambruysk commented on internal/dbstorage/dbstorage.go:
```diff
@@ -0,0 +1,144 @@
+
+//Ping: returns true if db is avilible
+func (st *storage) Ping(ctx context.Context) bool {
+	if err := st.db.PingContext(ctx); err != nil {
```
return st.db.PingContext(ctx) == nil ?

bambruysk commented on internal/api/api_test.go:
```diff
@@ -54,7 +59,7 @@ func Test_api_CrShort(t *testing.T) {
 			}
-			shurl, err2 := storage.GetByLong(tt.want)
+			shurl, err2 := storage.GetByLong(tt.want, req.Context())
 			if err2 != nil {
```
assert.NoError(t,err)

bambruysk commented on internal/api/api_test.go:
```diff
@@ -37,7 +38,11 @@ func Test_api_CrShort(t *testing.T) {
 			usecase := idgen.New(storage)
-			api := New(logger, &Options{}, usecase, storage)
+			usermanager, err := usermanager.New()
+			if err != nil {
```
здесь не log а t должен быть, чтобы в логи тестов попало

bambruysk commented on internal/api/api.go:
```diff
@@ -218,12 +264,112 @@ func (a *api) mdwDecompression(c *gin.Context) {
+	}
+
+	input := []jsonBlkIn{}
+	raw := []jsonBlkIn{}
```
можно не оперелять - а вывести прям на ходу же

bambruysk commented on internal/api/api.go:
```diff
@@ -218,12 +264,112 @@ func (a *api) mdwDecompression(c *gin.Context) {
+
+//convertBulk: POST multiple urls to shorten at once
+func (a *api) convertBulk(c *gin.Context) {
+	defer c.Request.Body.Close()
```
логгирование!

bambruysk commented on internal/dbstorage/dbstorage.go:
```diff
@@ -0,0 +1,144 @@
+	for _, el := range(data) {
+		_, err := stmt.Exec(el.Short[len(el.Short)-8:], el.URL, el.CrlID)
+		if err != nil {
+			if err = tx.Rollback(); err != nil {
```
зачем роллбэк делать здесь если он в дефер?

bambruysk reviewed (commented):
Небольшие правки - в целом хорошо

icyrogue committed:
fixing after rev inc13

fixing user manager inc13

v2 fixing user manager inc13

api and tests inc13

fixing usermanager v3 inc13

bambruysk commented on internal/urlstorage/urlstorage.go:
```diff
@@ -85,7 +88,8 @@ func (st *storage) Add(id string, long string) error {
 
 //TODO: возможно стоит заставить и все остольные функции storage взаимодействовать с файлом, тогда можно не выгружать все в память, но, с другой стороны, у нас там дальше нужно будет базу данных добавить, поэтому я не знаю
-func (st *storage) GetByID(id string) (string, error) {
+func (st *storage) GetByID(id string, ctx context.Context) (string, error) {
```
контекст принято предавать первым аргументом. если он не нужен то можно записать так func A( _ context.Context, b int) error - и сразу все понятно и строка не занимается.

bambruysk commented on internal/idgen/genid.go:
```diff
@@ -34,10 +37,32 @@ func (u *usecase) CreateShortURL(long string) (shurl string, err error) {
+		el.Short = (baseURL + "/" + genID())
+		log.Println(el)
+	}
+	/*Cкорей всго правильнее это сделать, добавляя новый эле-
```
в целом это может быть слайс указателей - тогда можно проще бегать по нему и менять его содержимое,
вообще если нужно создать новый слайс из дургого то идиоматично так
newSlice := make([]int,0, len(oldSlice)) // память выделяется один раз
for _,a := range oldSlice {
    newSlice = append(newSlice, a*2) // пишем в уже выделенную память
}

bambruysk commented on internal/idgen/genid.go:
```diff
@@ -34,10 +37,32 @@ func (u *usecase) CreateShortURL(long string) (shurl string, err error) {
+
+//BulkCreation: create multiple ids and append to database
+func (u usecase) BulkCreation(data []jsonmodels.JSONBulkInput, baseURL string) ([]jsonmodels.JSONBulkInput, error) {
+	for i := range(data){
```
скобки не нужны

bambruysk commented on internal/dbstorage/dbstorage.go:
```diff
@@ -0,0 +1,137 @@
+		return err
+	}
+	defer tx.Rollback()
+	stmt, err := tx.Prepare("INSERT INTO urls(id, long, token) VALUES($1, $2, $3);")
```
а если конфликт?

icyrogue replied:
Если в базе данных уже есть такая строчка, то такой запрос не доходит до  storage. А какой еще моет быть конфликт?

bambruysk reviewed (approved):
В целом неплохо, даже очень Можно сливать
